hub/core/storage/gcs.py

```diff
@@ -0,0 +1,277 @@
+
+        Example:
+            gcs_provider = GCSProvider("gcs://my-bucket/gcs_ds")
+
+        Args:
+            root (str): The root of the provider. All read/write request keys will be appended to root.
+            token (str/Dict): GCP token, used for fetching credentials for storage).
+                Can be a path to the credentials file, actual credential dictionary or one of the folowing:
+                `google_default`: Tries to load default credentials for the specified project.
+                `cache`: Retrieves the previously used credentials from cache if exist.
+                `anon`: Sets credentials=None
+                `browser`: Generates and stores new token file using cli.
+            project (str): Name of the project from gcloud.
+        """
+        self.root = root
+        self.token: Union[str, Dict, None] = token
+        self.project = project
+        self.missing_exceptions = (
+            FileNotFoundError,
+            IsADirectoryError,
+            NotADirectoryError,
+            AttributeError,
+        )
+        self._initialize_provider()
+
+    def _initialize_provider(self):
+        self._set_bucket_and_path()
+        if not self.token:
+            self.token = None
+        self.scoped_credentials = GCloudCredentials(self.token, project=self.project)
+        self.retry = retry.Retry(deadline=60)
+        client = storage.Client(credentials=self.scoped_credentials.credentials)
+        self.client_bucket = client.get_bucket(self.bucket)
+
+    def _reinitialize_provider(self):
+        client = storage.Client(credentials=self.scoped_credentials.credentials)
+        self.client_bucket = client.get_bucket(self.bucket)
+
+    def _set_bucket_and_path(self):
+        root = self.root.replace("gcp://", "").replace("gcs://", "")
+        self.bucket = root.split("/")[0]
+        self.path = root
+        if not self.path.endswith("/"):
+            self.path += "/"
+
+    def _get_path_from_key(self, key):
+        return posixpath.join(self.path, key)
+
+    def _list_keys(self):
+        self._blob_objects = self.client_bucket.list_blobs(prefix=self.path)
+        return [obj.name for obj in self._blob_objects]
+
+    def clear(self):
+        """Remove all keys below root - empties out mapping"""
+        self.check_readonly()
+        blob_objects = self.client_bucket.list_blobs(prefix=self.path)
+        for blob in blob_objects:
+            blob.delete()
+
+    def __getitem__(self, key):
+        """Retrieve data"""
+        try:
+            blob = self.client_bucket.get_blob(self._get_path_from_key(key))
+            return blob.download_as_bytes(retry=self.retry)
+        except self.missing_exceptions:
+            raise KeyError(key)
+
+    def __setitem__(self, key, value):
+        """Store value in key"""
+        self.check_readonly()
+        blob = self.client_bucket.blob(self._get_path_from_key(key))
+        if isinstance(value, memoryview):
```

Comment:
is there a way to do this without the memcopy? Might speed things up

Reply:
```python
            if isinstance(value, memoryview):
                if value.strides == (1,) and value.shape == (len(value.obj),):
                    value = value.obj
                else:
                    value = bytes(value)
```